File: web/lib/terms.ts
// Rolling internship terms, mirrored from src/terms.py so the Settings >
// Watch page can preview a lead-time or horizon change before the watcher
// runs. Same season start dates, same window arithmetic, same statuses;
// keep the two in step (tests on both sides pin the same dates).

export const SEASONS = ["Spring", "Summer", "Fall"] as const;
export type Season = (typeof SEASONS)[number];

const SEASON_START: Record<string, [number, number]> = {
  Spring: [1, 10],
  Summer: [5, 20],
  Fall: [8, 20],
  Winter: [12, 1],
};

export const DEFAULT_LEAD_WEEKS = 3;
export const DEFAULT_HORIZON_MONTHS = 14;

const TERM_RE = /^\s*(Spring|Summer|Fall|Winter)\s+(20\d\d)\s*$/i;

/** Calendar date without a time zone: "YYYY-MM-DD" in, plain arithmetic out. */
export type Day = { y: number; m: number; d: number };

export function day(y: number, m: number, d: number): Day {
  return { y, m, d };
}

export function dayFromIso(iso: string): Day {
  const [y, m, d] = iso.split("-").map(Number);
  return { y, m, d };
}

export function isoOf(a: Day): string {
  return `${a.y}-${String(a.m).padStart(2, "0")}-${String(a.d).padStart(2, "0")}`;
}

/** Today in the browser's local calendar. */
export function todayLocal(now: Date = new Date()): Day {
  return { y: now.getFullYear(), m: now.getMonth() + 1, d: now.getDate() };
}

function toUtc(a: Day): number {
  return Date.UTC(a.y, a.m - 1, a.d);
}

function fromUtc(ms: number): Day {
  const t = new Date(ms);
  return { y: t.getUTCFullYear(), m: t.getUTCMonth() + 1, d: t.getUTCDate() };
}

export function compare(a: Day, b: Day): number {
  return toUtc(a) - toUtc(b);
}

export function addDays(a: Day, n: number): Day {
  return fromUtc(toUtc(a) + n * 86_400_000);
}

/** Calendar-month arithmetic, clamping the day (Jan 31 + 1 -> Feb 28). */
export function addMonths(a: Day, months: number): Day {
  const month0 = a.m - 1 + months;
  const y = a.y + Math.floor(month0 / 12);
  const m = ((month0 % 12) + 12) % 12 + 1;
  const last = new Date(Date.UTC(y, m, 0)).getUTCDate();
  return { y, m, d: Math.min(a.d, last) };
}

export function parseTerm(term: string): { season: string; year: number } | null {
  const m = TERM_RE.exec(term ?? "");
  if (!m) return null;
  const season = m[1][0].toUpperCase() + m[1].slice(1).toLowerCase();
  return { season, year: Number(m[2]) };
}

export function termStart(term: string): Day | null {
  const p = parseTerm(term);
  if (!p) return null;
  const [m, d] = SEASON_START[p.season];
  return { y: p.year, m, d };
}

export function termSeason(term: string): Season | null {
  const p = parseTerm(term);
  return p && (SEASONS as readonly string[]).includes(p.season) ? (p.season as Season) : null;
}

export function window(today: Day, leadWeeks: number, horizonMonths: number): [Day, Day] {
  return [addDays(today, leadWeeks * 7), addMonths(today, horizonMonths)];
}

export function rollingTerms(
  today: Day,
  leadWeeks = DEFAULT_LEAD_WEEKS,
  horizonMonths = DEFAULT_HORIZON_MONTHS,
): string[] {
  const [lo, hi] = window(today, leadWeeks, horizonMonths);
  const out: string[] = [];
  for (let y = lo.y; y <= hi.y; y++) {
    for (const season of SEASONS) {
      const [m, d] = SEASON_START[season];
      const start = { y, m, d };
      if (compare(lo, start) <= 0 && compare(start, hi) <= 0) out.push(`${season} ${y}`);
    }
  }
  return out;
}

export function sortTerms(terms: string[]): string[] {
  const known = terms.filter((t) => termStart(t));
  const unknown = terms.filter((t) => !termStart(t));
  known.sort((a, b) => compare(termStart(a)!, termStart(b)!));
  return [...known, ...unknown];
}

export type TermStatus = "auto" | "included" | "excluded" | "past" | "beyond";

export type TermRow = {
  term: string;
  start: string;
  wanted: boolean;
  status: TermStatus;
  addedOn: string;
  dropsOn: string;
};

export type TermsConfig = {
  leadWeeks: number;
  horizonMonths: number;
  include: string[];
  exclude: string[];
};

/**
 * Every term worth showing: the window's terms, the pinned ones, and the
 * term that most recently dropped out (so "Fall 2026 · dropped Aug 3"
 * explains its absence). Chronological. Mirrors src/terms.py term_rows.
 */
export function termRows(cfg: TermsConfig, today: Day): TermRow[] {
  const include = cfg.include.filter(parseTerm);
  const exclude = cfg.exclude.filter(parseTerm);
  const [lo] = window(today, cfg.leadWeeks, cfg.horizonMonths);
  const auto = rollingTerms(today, cfg.leadWeeks, cfg.horizonMonths);

  let previous: string | null = null;
  for (let y = lo.y - 1; y <= lo.y; y++) {
    for (const season of SEASONS) {
      const [m, d] = SEASON_START[season];
      if (compare({ y, m, d }, lo) < 0) previous = `${season} ${y}`;
    }
  }

  const names = sortTerms([...new Set([...auto, ...include, ...exclude, ...(previous ? [previous] : [])])]);
  return names.map((term) => {
    const start = termStart(term)!;
    let status: TermStatus;
    let wanted: boolean;
    if (exclude.includes(term)) [status, wanted] = ["excluded", false];
    else if (include.includes(term)) [status, wanted] = ["included", true];
    else if (auto.includes(term)) [status, wanted] = ["auto", true];
    else if (compare(start, lo) < 0) [status, wanted] = ["past", false];
    else [status, wanted] = ["beyond", false];
    return {
      term,
      start: isoOf(start),
      wanted,
      status,
      addedOn: isoOf(addMonths(start, -cfg.horizonMonths)),
      dropsOn: isoOf(addDays(start, -cfg.leadWeeks * 7)),
    };
  });
}

/** "Aug 3" / "Aug 3, 2027" - short date for the term rows. */
export function shortDate(iso: string, today: Day): string {
  const d = dayFromIso(iso);
  const month = new Date(Date.UTC(d.y, d.m - 1, d.d)).toLocaleString("en-US", {
    month: "short",
    timeZone: "UTC",
  });
  return d.y === today.y ? `${month} ${d.d}` : `${month} ${d.d}, ${d.y}`;
}

File: src/terms.py
"""Rolling internship terms.

A static `terms_wanted` list rots: nobody remembers to drop "Fall 2026" in
August or to add "Fall 2027" in June. With `terms:` in the user yaml the
wanted set is derived from today's date instead: every Spring/Summer/Fall
term whose start lies between `lead_weeks` from now and `horizon_months`
from now is wanted, and `include` / `exclude` pin explicit exceptions.

    terms:
      rolling: true
      lead_weeks: 3          # stop wanting a term this close to its start
      horizon_months: 14     # don't want a term starting further out
      include: []            # always wanted, e.g. ["Summer 2028"]
      exclude: []            # never wanted, e.g. ["Spring 2027"]

The legacy `terms_wanted: [...]` list keeps working unchanged when `terms:`
is absent. `web/lib/terms.ts` mirrors the window arithmetic so the settings
page can preview the effect of a lead-time change before the watcher runs;
keep the two in step.
"""

from __future__ import annotations

import datetime as dt
import re
from dataclasses import asdict, dataclass

# Seasons the rolling window generates, chronological within a year. Winter
# is parseable (an `include` may name one) but never generated: winter
# co-ops are rare and would mostly add noise.
SEASONS = ("Spring", "Summer", "Fall")
SEASON_START: dict[str, tuple[int, int]] = {
    "Spring": (1, 10), "Summer": (5, 20), "Fall": (8, 20), "Winter": (12, 1),
}
DEFAULT_LEAD_WEEKS = 3
DEFAULT_HORIZON_MONTHS = 14

_TERM_RE = re.compile(r"^\s*(Spring|Summer|Fall|Winter)\s+(20\d\d)\s*$", re.I)


def parse_term(term: str) -> tuple[str, int] | None:
    """'Fall 2026' -> ('Fall', 2026); None when it isn't a season + year."""
    m = _TERM_RE.match(term or "")
    if not m:
        return None
    return m.group(1).title(), int(m.group(2))


def term_start(term: str) -> dt.date | None:
    parsed = parse_term(term)
    if parsed is None:
        return None
    season, year = parsed
    month, day = SEASON_START[season]
    return dt.date(year, month, day)


def term_season(term: str) -> str | None:
    parsed = parse_term(term)
    return parsed[0] if parsed else None


def add_months(day: dt.date, months: int) -> dt.date:
    """Calendar-month arithmetic, clamping the day (Jan 31 + 1 -> Feb 28)."""
    month0 = day.month - 1 + months
    year = day.year + month0 // 12
    month = month0 % 12 + 1
    last = (dt.date(year + (month == 12), month % 12 + 1, 1)
            - dt.timedelta(days=1)).day
    return dt.date(year, month, min(day.day, last))


def window(today: dt.date, lead_weeks: int,
           horizon_months: int) -> tuple[dt.date, dt.date]:
    """(earliest, latest) term start the rolling window wants today."""
    return (today + dt.timedelta(weeks=lead_weeks),
            add_months(today, horizon_months))


def rolling_terms(today: dt.date, lead_weeks: int = DEFAULT_LEAD_WEEKS,
                  horizon_months: int = DEFAULT_HORIZON_MONTHS) -> list[str]:
    """Generated terms inside the window, chronological."""
    lo, hi = window(today, lead_weeks, horizon_months)
    out: list[str] = []
    for year in range(lo.year, hi.year + 1):
        for season in SEASONS:
            month, day = SEASON_START[season]
            if lo <= dt.date(year, month, day) <= hi:
                out.append(f"{season} {year}")
    return out


def sort_terms(terms: list[str]) -> list[str]:
    """Chronological; unparseable terms keep their relative order at the end."""
    known = [t for t in terms if term_start(t) is not None]
    unknown = [t for t in terms if term_start(t) is None]
    known.sort(key=lambda t: term_start(t) or dt.date.max)
    return known + unknown


@dataclass
class TermRow:
    """One term as the settings page shows it: wanted or not, and why."""
    term: str
    start: str                 # ISO date
    wanted: bool
    # auto: inside the window | included / excluded: pinned by config |
    # past: start is closer than lead_weeks (or gone) | beyond: past horizon
    status: str
    added_on: str              # ISO date the window first wants it
    drops_on: str              # ISO date the window stops wanting it


def canonical(term: str) -> str | None:
    """'  summer 2027 ' -> 'Summer 2027', the spelling generated terms and
    job.terms use; None when it isn't a term."""
    parsed = parse_term(term)
    return f"{parsed[0]} {parsed[1]}" if parsed else None


def canonical_list(terms: list[str] | None) -> list[str]:
    """Canonical spellings, unparseable entries dropped, order kept, no dups."""
    out: list[str] = []
    for raw in terms or []:
        c = canonical(str(raw))
        if c and c not in out:
            out.append(c)
    return out


def _config(cfg_terms: dict | None) -> tuple[int, int, list[str], list[str]]:
    cfg_terms = cfg_terms or {}
    lead = int(cfg_terms.get("lead_weeks", DEFAULT_LEAD_WEEKS))
    horizon = int(cfg_terms.get("horizon_months", DEFAULT_HORIZON_MONTHS))
    return (lead, horizon, canonical_list(cfg_terms.get("include")),
            canonical_list(cfg_terms.get("exclude")))


def term_rows(cfg_terms: dict | None, today: dt.date) -> list[TermRow]:
    """Every term worth showing: the window's terms, the pinned ones, and
    the term that most recently dropped out (so "Fall 2026: dropped Aug 3"
    explains its absence). Chronological."""
    lead, horizon, include, exclude = _config(cfg_terms)
    lo, hi = window(today, lead, horizon)
    auto = rolling_terms(today, lead, horizon)

    # The most recent generated term whose start fell before the window.
    previous: str | None = None
    for year in range(lo.year - 1, lo.year + 1):
        for season in SEASONS:
            month, day = SEASON_START[season]
            if dt.date(year, month, day) < lo:
                previous = f"{season} {year}"

    names = sort_terms(list(dict.fromkeys(
        auto + include + exclude + ([previous] if previous else []))))
    rows: list[TermRow] = []
    for term in names:
        start = term_start(term)
        assert start is not None  # everything above parsed
        if term in exclude:
            status, wanted = "excluded", False
        elif term in include:
            status, wanted = "included", True
        elif term in auto:
            status, wanted = "auto", True
        elif start < lo:
            status, wanted = "past", False
        else:
            status, wanted = "beyond", False
        rows.append(TermRow(
            term=term, start=start.isoformat(), wanted=wanted, status=status,
            added_on=add_months(start, -horizon).isoformat(),
            drops_on=(start - dt.timedelta(weeks=lead)).isoformat()))
    return rows


def wanted_terms(cfg: dict, today: dt.date) -> list[str]:
    """The user's wanted terms today, chronological. `terms:` (rolling)
    wins over the legacy `terms_wanted:` list when both are present."""
    cfg_terms = cfg.get("terms")
    if isinstance(cfg_terms, dict) and cfg_terms.get("rolling", True):
        return [r.term for r in term_rows(cfg_terms, today) if r.wanted]
    # Legacy list: canonical spellings where they parse, so "fall 2026"
    # still meets a job tagged "Fall 2026"; anything else passes through
    # untouched (a user may list a term shape this module doesn't know).
    legacy = [canonical(str(t)) or str(t)
              for t in cfg.get("terms_wanted") or [] if t]
    return sort_terms(list(dict.fromkeys(legacy)))


def rows_as_dicts(rows: list[TermRow]) -> list[dict]:
    return [asdict(r) for r in rows]
